templates/accounts.go

```diff
@@ -26,58 +26,126 @@ import (
 )
 
 const createAccountTemplate = `
-transaction(publicKeys: [[UInt8]], code: [UInt8]) {
+transaction(publicKeys: [[UInt8]]) {
   prepare(signer: AuthAccount) {
 	let acct = AuthAccount(payer: signer)
 
 	for key in publicKeys {
 	  acct.addPublicKey(key)
 	}
-	
-	if code.length > 0 {
-	  acct.setCode(code)
+  }
+}
+`
+
+const createAccountWitContractsTemplate = `
+transaction(publicKeys: [[UInt8]], contracts: {String: [UInt8]}) {
+  prepare(signer: AuthAccount) {
+	let acct = AuthAccount(payer: signer)
+
+	for key in publicKeys {
+	  acct.addPublicKey(key)
+	}
+
+    for contract in contracts.keys {
+	  acct.contracts.add(name: contract, code: contracts[contract]!)
 	}
   }
 }
 `
 
 // CreateAccount generates a transactions that creates a new account.
 //
-// This template accepts a list of public keys and a code argument, both of which are optional.
+// This template accepts an optional list of public keys.
 //
 // The final argument is the address of the account that will pay the account creation fee.
 // This account is added as a transaction authorizer and therefore must sign the resulting transaction.
-func CreateAccount(accountKeys []*flow.AccountKey, code []byte, payer flow.Address) *flow.Transaction {
+func CreateAccount(accountKeys []*flow.AccountKey, payer flow.Address) *flow.Transaction {
 	publicKeys := make([]cadence.Value, len(accountKeys))
 
 	for i, accountKey := range accountKeys {
 		publicKeys[i] = bytesToCadenceArray(accountKey.Encode())
 	}
 
 	cadencePublicKeys := cadence.NewArray(publicKeys)
-	cadenceCode := bytesToCadenceArray(code)
 
 	return flow.NewTransaction().
 		SetScript([]byte(createAccountTemplate)).
 		AddAuthorizer(payer).
+		AddRawArgument(jsoncdc.MustEncode(cadencePublicKeys))
+}
+
+// CreateAccountWithContracts generates a transactions that creates a new account.
+//
+// This template accepts a list of public keys and a contracts argument, both of which are optional.
+//
+// The contracts argument is a dictionary of *contract name*: *contract code (in bytes)*.
+// All of the contracts will be deployed to the account.
+//
+// The final argument is the address of the account that will pay the account creation fee.
+// This account is added as a transaction authorizer and therefore must sign the resulting transaction.
+func CreateAccountWithContracts(accountKeys []*flow.AccountKey, contracts map[string][]byte, payer flow.Address) *flow.Transaction {
+	publicKeys := make([]cadence.Value, len(accountKeys))
+
+	for i, accountKey := range accountKeys {
+		publicKeys[i] = bytesToCadenceArray(accountKey.Encode())
+	}
+
+	contractKeyPairs := make([]cadence.KeyValuePair, len(contracts))
+	i := 0
+	for s, bytes := range contracts {
```

Comment:
Map iteration order in Go is non-deterministic, so this function may generate different Cadence argument orderings on subsequent runs with the same input.

This probably isn't a problem in most cases, but could be confusing for users who expect deterministic transaction generation in their tests. 

What if we added a dedicated type for contracts? It could even be defined in the `cadence` package:

```go
type Contract struct {
  Name string
  Source []byte
}
```

```go
import "github.com/onflow/cadence"

func CreateAccountWithContracts(accountKeys []*flow.AccountKey, contracts []cadence.Contract, payer flow.Address) *flow.Transaction {
  // ... 
}

contracts := []cadence.Contract{
  cadence.NewContract("MyContract", myContractCode),
}

tx := CreateAccountWithContracts(keys, contracts, payer)
```

@turbolent @janezpodhostnik

Reply:
Good point Peter! The recommendation looks good

Reply:
@turbolent @psiemens switched to list. I double checked there isn't a similar issue in flow-go.